Untuk export data ke google colabs dari google drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


import modul

In [ ]:
import xarray as xr
import numpy as np
import pandas as pd
import tensorflow as tf
from datetime import date
from google.colab import files
from google.colab import drive
import shutil
import csv

In [ ]:
import matplotlib.pyplot as plt

# Pengolahan data

In [ ]:
data  = xr.open_dataset('/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_'+"%04d" % (2020)+"%02d" % (12)+"%02d" % (1)+'_vDT2021.nc')
data  = data.where((data['latitude'] > -13.65) & (data['latitude'] < 9.5) & (data['longitude'] > 93.6) & (data['longitude'] < 142.7), drop=True)
data

<xarray.Dataset>
Dimensions:         (latitude: 93, longitude: 197, nv: 2, time: 1)
Coordinates:
  * time            (time) datetime64[ns] 2020-12-01
  * latitude        (latitude) float32 -13.62 -13.38 -13.12 ... 9.125 9.375
  * longitude       (longitude) float32 93.62 93.88 94.12 ... 142.1 142.4 142.6
  * nv              (nv) int32 0 1
Data variables: (12/14)
    crs             (latitude, longitude) float64 -2.147e+09 ... -2.147e+09
    lat_bnds        (latitude, nv, longitude) float32 -13.75 -13.75 ... 9.5 9.5
    lon_bnds        (longitude, nv, latitude) float32 93.5 93.5 ... 142.8 142.8
    sla             (time, latitude, longitude) float64 0.1868 0.1926 ... 0.1445
    err_sla         (time, latitude, longitude) float64 0.0144 0.0146 ... 0.0137
    ugosa           (time, latitude, longitude) float64 0.0424 0.0653 ... 0.1591
    ...              ...
    err_vgosa       (time, latitude, longitude) float64 0.0565 0.0544 ... 0.0727
    adt             (time, latitude, longitude) float64 1.111 1.118 ... 1.14
    ugos            (time, latitude, longitude) float64 0.01 0.0332 ... -0.0112
    vgos            (time, latitude, longitude) float64 -0.0608 ... -0.017
    flag_ice        (time, latitude, longitude) float64 0.0 0.0 0.0 ... 0.0 0.0
    tpa_correction  (time, latitude, longitude) float64 0.0 0.0 0.0 ... 0.0 0.0
Attributes: (12/44)
    Conventions:                     CF-1.6
    Metadata_Conventions:            Unidata Dataset Discovery v1.0
    cdm_data_type:                   Grid
    comment:                         Sea Surface Height measured by Altimetry...
    contact:                         http://climate.copernicus.eu/c3s-user-se...
    creator_email:                   http://climate.copernicus.eu/c3s-user-se...
    ...                              ...
    summary:                         SSALTO/DUACS Delayed-Time Level-4 sea su...
    time_coverage_duration:          P1D
    time_coverage_end:               2020-12-01T12:00:00Z
    time_coverage_resolution:        P1D
    time_coverage_start:             2020-11-30T12:00:00Z
    title:                           DT merged two satellites Global Ocean Gr...

In [ ]:
import pickle
with open('coor.pickle','wb') as coordinate:
  pickle.dump(coor,coordinate,protocol=pickle.HIGHEST_PROTOCOL)

In [ ]:
coor = {'lat':data['lat_bnds'].values.mean(axis=1)[1:-1,1:-1],
        'lon':data['lon_bnds'].values.mean(axis=1)[1:-1,1:-1].T}

coor

{'lat': array([[-13.375, -13.375, -13.375, ..., -13.375, -13.375, -13.375],
        [-13.125, -13.125, -13.125, ..., -13.125, -13.125, -13.125],
        [-12.875, -12.875, -12.875, ..., -12.875, -12.875, -12.875],
        ...,
        [  8.625,   8.625,   8.625, ...,   8.625,   8.625,   8.625],
        [  8.875,   8.875,   8.875, ...,   8.875,   8.875,   8.875],
        [  9.125,   9.125,   9.125, ...,   9.125,   9.125,   9.125]],
       dtype=float32),
 'lon': array([[ 93.875,  94.125,  94.375, ..., 141.875, 142.125, 142.375],
        [ 93.875,  94.125,  94.375, ..., 141.875, 142.125, 142.375],
        [ 93.875,  94.125,  94.375, ..., 141.875, 142.125, 142.375],
        ...,
        [ 93.875,  94.125,  94.375, ..., 141.875, 142.125, 142.375],
        [ 93.875,  94.125,  94.375, ..., 141.875, 142.125, 142.375],
        [ 93.875,  94.125,  94.375, ..., 141.875, 142.125, 142.375]],
       dtype=float32)}

In [ ]:
Lat = np.empty(197)
Lat[:] = np.nan
Lon = np.empty(93)
Lon[:] = np.nan
Lat2 = np.empty(197*2)
Lat2[:] = np.nan
Lon2 = np.empty(93*2)
Lon2[:] = np.nan

#diagonal
latd = np.empty(93)
latd[:] = np.nan
latd=latd.reshape(93,1)

lond = np.empty(197+1)
lond[:] = np.nan
lond=lond.reshape(1,197+1)

num_nan = np.empty(197*93)
num_nan[:] = np.nan

data train

In [ ]:
#Olah data train
center  = []
south   = []
north   = []
east    = []
west    = []

u_geos  = []
v_geos  = []
for year in range(2013,2022): #nanti 2019 2022
  for month in range(1,13):
    for day in range(1,32):
      try:
        #ambil data nc
        data  = xr.open_dataset('/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_'+"%04d" % (year)+"%02d" % (month)+"%02d" % (day)+'_vDT2021.nc')
        data  = data.where((data['latitude'] > -13.65) & (data['latitude'] < 9.5) & (data['longitude'] > 93.6) & (data['longitude'] < 142.7), drop=True)


        center.append(data['adt'].to_numpy().reshape(93,197)[1:-1,1:-1])
        south.append(np.roll(data['adt'].to_numpy(), 1,axis=1).reshape(93,197)[1:-1,1:-1])
        north.append(np.roll(data['adt'].to_numpy(),-1,axis=1).reshape(93,197)[1:-1,1:-1])
        west.append(np.roll(data['adt'].to_numpy(), 1,axis=2).reshape(93,197)[1:-1,1:-1])
        east.append(np.roll(data['adt'].to_numpy(),-1,axis=2).reshape(93,197)[1:-1,1:-1])

        u_geos.append(data['ugos'].to_numpy().reshape(93,197)[1:-1,1:-1])
        v_geos.append(data['vgos'].to_numpy().reshape(93,197)[1:-1,1:-1])

      except:
        print('No such file or directory for year :'+str(year)+" month :"+str(month)+" day :"+str(day))
        break

No such file or directory for year :2013 month :2 day :29
No such file or directory for year :2013 month :4 day :31
No such file or directory for year :2013 month :6 day :31
No such file or directory for year :2013 month :9 day :31
No such file or directory for year :2013 month :11 day :31
No such file or directory for year :2014 month :2 day :29
No such file or directory for year :2014 month :4 day :31
No such file or directory for year :2014 month :6 day :31
No such file or directory for year :2014 month :9 day :31
No such file or directory for year :2014 month :11 day :31
No such file or directory for year :2015 month :2 day :29
No such file or directory for year :2015 month :4 day :31
No such file or directory for year :2015 month :6 day :31
No such file or directory for year :2015 month :9 day :31
No such file or directory for year :2015 month :11 day :31
No such file or directory for year :2016 month :2 day :30
No such file or directory for year :2016 month :4 day :31
No such fil

In [ ]:
import pickle
with open('/content/drive/MyDrive/Tugas Akhir/raw_data.pickle','wb') as w:
  pickle.dump({'center':center,
          'south':south,
          'north':north,
          'east':east,
          'west':west,
          'u_geos':u_geos,
          'v_geos':v_geos}, w, protocol=pickle.HIGHEST_PROTOCOL)


In [ ]:
{'center':center,
          'south':south,
          'north':north,
          'east':east,
          'west':west,
          'u_geos':u_geos,
          'v_geos':v_geos}

{'center': [array([[1.136 , 1.117 , 1.0997, ...,    nan,    nan,    nan],
         [1.1161, 1.0992, 1.0857, ...,    nan,    nan,    nan],
         [1.0966, 1.0831, 1.0705, ...,    nan,    nan,    nan],
         ...,
         [1.0145, 1.0277, 1.0425, ..., 1.0969, 1.0974, 1.0938],
         [1.0044, 1.0155, 1.0297, ..., 1.1124, 1.1125, 1.1139],
         [0.9975, 1.0037, 1.0186, ..., 1.1217, 1.1255, 1.1278]]),
  array([[1.1172, 1.1   , 1.0844, ...,    nan,    nan,    nan],
         [1.1019, 1.0904, 1.0762, ...,    nan,    nan,    nan],
         [1.0849, 1.0727, 1.0612, ...,    nan,    nan,    nan],
         ...,
         [1.0156, 1.0283, 1.0424, ..., 1.0959, 1.0938, 1.0945],
         [1.0052, 1.017 , 1.0285, ..., 1.1095, 1.1093, 1.1087],
         [0.9999, 1.0097, 1.0208, ..., 1.1221, 1.1213, 1.1222]]),
  array([[1.101 , 1.0853, 1.0714, ...,    nan,    nan,    nan],
         [1.086 , 1.0723, 1.0618, ...,    nan,    nan,    nan],
         [1.071 , 1.0584, 1.05  , ...,    nan,    nan,    nan]

In [ ]:
#simpan data nc ke csv

with open("Data_train.csv", "w") as outfile:

    writer = csv.writer(outfile)

    key_list = list(X.keys())

    limit = len(X['ADT_C'])

    writer.writerow(X.keys())

    for i in range(limit):
        writer.writerow([X[x][i] for x in key_list])

In [ ]:
#simpan data train ke csv

shutil.copy("/content/Data_train.csv","/content/drive/MyDrive/Tugas Akhir")

In [ ]:
#hilangkan data nan di train
Data_Train=pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Data_train.csv')
Data_Train=Data_Train.dropna()
Data_Train.to_csv('Data_train_fix.csv')

shutil.copy("/content/Data_train_fix.csv","/content/drive/MyDrive/Tugas Akhir")

data test

In [ ]:
#Olah data test

X={
       "ADT_C": np.NaN,
       "ADT_N": np.NaN,
       "ADT_S": np.NaN,
       "ADT_W": np.NaN,
       "ADT_E": np.NaN,

       "lat"   : np.NaN,
       "lon"   : np.NaN,

       "Ugos" : np.NaN,
       "Vgos" : np.NaN,
       "U"    : np.NaN,
       "V"    : np.NaN
       }
for year in range(2019,2022):
  for month in range(1,13):
    for day in range(1,32):
      try:
        print("/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_"+"%04d"%(year)+"%02d" % (month)+"%02d" % (day)+"_vDT2021.nc")
        Data  = xr.open_dataset("/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_"+"%04d"%(year)+"%02d" % (month)+"%02d" % (day)+"_vDT2021.nc")


        #ADT stensil
        ADT_xy= Data.adt[0,305:398,1094:1291].to_numpy()

        #1st stensil
        X["ADT_S"]  = np.append(X["ADT_S"],np.concatenate((Lat, ADT_xy), axis=None)[:-197].reshape((1,-1))[0])
        X["ADT_N"]  = np.append(X["ADT_N"],np.concatenate((ADT_xy,Lat), axis=None)[197:].reshape((1,-1))[0])
        X["ADT_W"]  = np.append(
            X["ADT_W"],np.concatenate((Lon,ADT_xy.transpose()), axis=None).reshape(
                (-1,93))[:-1,:].transpose().reshape((1,-1)).reshape((1,-1))[:])
        X["ADT_E"]  = np.append(
            X["ADT_E"],np.concatenate((ADT_xy.transpose(),Lon), axis=None).reshape(
                (-1,93))[1:,:].transpose().reshape((1,-1)).reshape((1,-1))[:])
        X["ADT_C"]  = np.append(X["ADT_C"],ADT_xy.reshape((1,-1))[0])

        X["Ugos"]   = np.append(X["Ugos"],Data.ugos[0,305:398,1094:1291].to_numpy().reshape((1,-1))[0])
        X["Vgos"]   = np.append(X["Vgos"],Data.vgos[0,305:398,1094:1291].to_numpy().reshape((1,-1))[0])
        X["U"]      = np.append(X["U"],Data.ugos[0,305:398,1094:1291].to_numpy().reshape((1,-1))[0])
        X["V"]      = np.append(X["V"],Data.vgos[0,305:398,1094:1291].to_numpy().reshape((1,-1))[0])
      except:
        print('No such file or directory for year :'+str(year)+" month :"+str(month)+" day :"+str(day))
        break
  #write
  #write
X["Ugos"]   = np.append(X["Ugos"],num_nan)[18321:]
X["Vgos"]   = np.append(X["Vgos"],num_nan)[18321:]

/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190101_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190102_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190103_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190104_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190105_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190106_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190107_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190108_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190109_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190110_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/dt_global_twosat_phy_l4_20190111_vDT2021.nc
/content/drive/MyDrive/Tugas Akhir/DATAFIX/

In [ ]:
with open("Data_train_2022.csv", "w") as outfile:

   # pass the csv file to csv.writer.
    writer = csv.writer(outfile)

    # convert the dictionary keys to a list
    key_list = list(X.keys())

    # find the length of the key_list
    limit = len(X['ADT_C'])

    # the length of the keys corresponds to
    # no. of. columns.
    writer.writerow(X.keys())

    # iterate each column and assign the
    # corresponding values to the column
    for i in range(limit):
        writer.writerow([X[x][i] for x in key_list])


In [ ]:
shutil.copy("/content/Data_train_2022.csv","/content/drive/MyDrive/Tugas Akhir")

'/content/drive/MyDrive/Tugas Akhir/Data_train_2022.csv'

In [ ]:
#hilangkan data nan di test
Data_Test  = pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Data_test.csv')
Data_Test  = Data_Test.dropna()
Data_Test.to_csv('Data_test_fix.csv')

shutil.copy("/content/Data_test_fix.csv","/content/drive/MyDrive/Tugas Akhir")

#Model

In [ ]:
import tensorflow as tf

In [ ]:
Data_Train = pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Data_test_12(SLA).csv',usecols=['ADT_C','ADT_N','ADT_S','ADT_W','ADT_E','U','V','Ugos','Vgos'])

#Data_Test  = pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Data_test_12(SLA).csv')

#Data_Train  = pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Data_train_12(SLA).csv')

In [ ]:
Data_Train.head()

ADT_C   ADT_N  ADT_S   ADT_W   ADT_E    Ugos    Vgos       U       V
1  1.0683  1.0761    NaN     NaN  1.0532  0.0250  0.1665  0.0893  0.1456
2  1.0532  1.0606    NaN  1.0683  1.0412 -0.0039  0.1033  0.0420  0.1563
3  1.0412  1.0483    NaN  1.0532  1.0374 -0.0125  0.0547  0.0113  0.0871
4  1.0374  1.0430    NaN  1.0412  1.0397 -0.0291 -0.0158 -0.0043 -0.0017
5  1.0397  1.0421    NaN  1.0374  1.0429 -0.0365 -0.0645 -0.0355 -0.0305

In [ ]:
Data_Train = Data_Train.drop(0)

In [ ]:
Data_Train = Data_Train.to_numpy()

In [ ]:
len(Data_Train)

40141311

In [ ]:
Data_Test

ADT_C   ADT_N   ADT_S   ADT_W   ADT_E   SLA_C   SLA_N   SLA_S  \
0            NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
1         1.0683  1.0761     NaN     NaN  1.0532  0.1444  0.1555     NaN   
2         1.0532  1.0606     NaN  1.0683  1.0412  0.1279  0.1385     NaN   
3         1.0412  1.0483     NaN  1.0532  1.0374  0.1144  0.1246     NaN   
4         1.0374  1.0430     NaN  1.0412  1.0397  0.1090  0.1178     NaN   
...          ...     ...     ...     ...     ...     ...     ...     ...   
20079812  1.1504     NaN  1.1509  1.1571  1.1465  0.1631     NaN  0.1737   
20079813  1.1465     NaN  1.1429  1.1504  1.1430  0.1575     NaN  0.1637   
20079814  1.1430     NaN  1.1375  1.1465  1.1397  0.1520     NaN  0.1568   
20079815  1.1397     NaN  1.1317  1.1430  1.1386  0.1463     NaN  0.1487   
20079816  1.1386     NaN  1.1310  1.1397     NaN  0.1433     NaN  0.1456   

           SLA_W   SLA_E     lat      lon    Ugos    Vgos       U       V  
0            NaN     NaN     NaN      NaN -0.0355 -0.0258     NaN     NaN  
1            NaN  0.1279 -13.625   93.625  0.0250  0.1665  0.0893  0.1456  
2         0.1444  0.1144 -13.625   93.875 -0.0039  0.1033  0.0420  0.1563  
3         0.1279  0.1090 -13.625   94.125 -0.0125  0.0547  0.0113  0.0871  
4         0.1144  0.1097 -13.625   94.375 -0.0291 -0.0158 -0.0043 -0.0017  
...          ...     ...     ...      ...     ...     ...     ...     ...  
20079812  0.1714  0.1575   9.375  141.625     NaN     NaN  0.0057 -0.1012  
20079813  0.1631  0.1520   9.375  141.875     NaN     NaN -0.0159 -0.0608  
20079814  0.1575  0.1463   9.375  142.125     NaN     NaN -0.0480 -0.0719  
20079815  0.1520  0.1433   9.375  142.375     NaN     NaN -0.0808 -0.0561  
20079816  0.1463     NaN   9.375  142.625     NaN     NaN -0.0995  0.0079  

[20079817 rows x 16 columns]

In [ ]:
Data_Test = Data_Test.drop(0)

NameError: ignored

#1 day prediction

In [ ]:
size=len(Data_Train)

Train_set=np.stack((
    Data_Train.ADT_C.to_numpy().reshape((size, 1)),
    Data_Train.ADT_N.to_numpy().reshape((size, 1)),
    Data_Train.ADT_E.to_numpy().reshape((size, 1)),
    Data_Train.ADT_W.to_numpy().reshape((size, 1)),
    Data_Train.ADT_S.to_numpy().reshape((size, 1)),
    Data_Train.V.to_numpy().reshape((size, 1)),
    Data_Train.U.to_numpy().reshape((size, 1)),
    Data_Train.Ugos.to_numpy().reshape((size, 1)),
    Data_Train.Vgos.to_numpy().reshape((size, 1))
    ),axis=1)
Train_set=Train_set.reshape((size, Train_set.shape[1]))
Train_set.shape

np.random.shuffle(Train_set)

X=Train_set[:,:-2]
Y=Train_set[:,-2:]

NameError: ignored

In [ ]:
size_test=len(Data_Test)

X_test=np.stack((
    Data_Test.ADT_C.to_numpy().reshape((size_test, 1)),
    Data_Test.ADT_N.to_numpy().reshape((size_test, 1)),
    Data_Test.ADT_E.to_numpy().reshape((size_test, 1)),
    Data_Test.ADT_W.to_numpy().reshape((size_test, 1)),
    Data_Test.ADT_S.to_numpy().reshape((size_test, 1)),

    Data_Test.V.to_numpy().reshape((size_test, 1)),
    Data_Test.U.to_numpy().reshape((size_test, 1))
    ),axis=1)
X_test=X_test.reshape((size_test, 7))

Y_test=np.stack((Data_Test.Ugos.to_numpy().reshape((size_test, 1)),
            Data_Test.Vgos.to_numpy().reshape((size_test, 1))),axis=1)
Y_test=Y_test.reshape((size_test, 2))

NameError: ignored

#general prediction


In [ ]:
size        = len(Data_Train)
Data_Train.ADT_C.to_numpy().reshape((size, 1)).shape

AttributeError: ignored

In [ ]:
def generate_data(Data,train=False,plot=False):

  #Data        = Data_Train.drop(0)

  size        = len(Data)

  num_nan2    = np.empty(197*93*1)
  num_nan2[:] = np.nan

  num_nan3    = np.empty(197*93*2)
  num_nan3[:] = np.nan

  num_nan4    = np.empty(197*93*3)
  num_nan4[:] = np.nan

  num_nan5    = np.empty(197*93*4)
  num_nan5[:] = np.nan

  #num_nan6    = np.empty(197*93*5)
  #num_nan6[:] = np.nan

  #num_nan7    = np.empty(197*93*6)
  #num_nan7[:] = np.nan

  #num_nan8    = np.empty(197*93*7)
  #num_nan8[:] = np.nan

  Set           = np.stack((

                          #Data.lat.to_numpy().reshape((size, 1)),
                          #Data.lon.to_numpy().reshape((size, 1)),
                          #LAT N
                          #Data.lat.to_numpy().reshape((size, 1))+0.25,
                          #LON W
                          #Data.lon.to_numpy().reshape((size, 1))-0.25,
                          #LAT S
                          #Data.lat.to_numpy().reshape((size, 1))-0.25,
                          #LON E
                          #Data.lon.to_numpy().reshape((size, 1))+0.25,
                          Data[:,0].reshape((size, 1)),
                          Data[:,1].reshape((size, 1)),
                          Data[:,2].reshape((size, 1)),
                          Data[:,3].reshape((size, 1)),
                          Data[:,4].reshape((size, 1)),
                          Data[:,7].reshape((size, 1)),
                          Data[:,8].reshape((size, 1)),
                          Data[:,5].reshape((size, 1)),
                          Data[:,6].reshape((size, 1)),
                          #d2
                          np.append(Data[:,5].reshape((size, 1)),num_nan2)[18321*1:].reshape((size, 1)),
                          np.append(Data[:,6].reshape((size, 1)),num_nan2)[18321*1:].reshape((size, 1)),
                          #d3
                          np.append(Data[:,5].reshape((size, 1)),num_nan3)[18321*2:].reshape((size, 1)),
                          np.append(Data[:,6].reshape((size, 1)),num_nan3)[18321*2:].reshape((size, 1)),
                          #d4
                          np.append(Data[:,5].reshape((size, 1)),num_nan4)[18321*3:].reshape((size, 1)),
                          np.append(Data[:,6].reshape((size, 1)),num_nan4)[18321*3:].reshape((size, 1)),
                          #d5
                          np.append(Data[:,5].reshape((size, 1)),num_nan5)[18321*4:].reshape((size, 1)),
                          np.append(Data[:,6].reshape((size, 1)),num_nan5)[18321*4:].reshape((size, 1)),

                          #d6
                          #np.append(Data.Ugos.to_numpy().reshape((size, 1)),num_nan6)[18321*5:].reshape((size, 1)),
                          #np.append(Data.Vgos.to_numpy().reshape((size, 1)),num_nan6)[18321*5:].reshape((size, 1)),

                          #d7
                          #np.append(Data.Ugos.to_numpy().reshape((size, 1)),num_nan7)[18321*6:].reshape((size, 1)),
                          #np.append(Data.Vgos.to_numpy().reshape((size, 1)),num_nan7)[18321*6:].reshape((size, 1)),

                          #d8
                          #np.append(Data.Ugos.to_numpy().reshape((size, 1)),num_nan8)[18321*7:].reshape((size, 1)),
                          #np.append(Data.Vgos.to_numpy().reshape((size, 1)),num_nan8)[18321*7:].reshape((size, 1))
                        ),
                          axis=1)
  Set           = Set.reshape((size, Set.shape[1]))
  #if train==True:
  if plot == False:
    Set           = Set[~np.isnan(Set).any(axis=1)]
  if train == True:
    np.random.shuffle(Set)

  #del Data
  return Set

(40141312, 10)

In [ ]:
def generate_data(Data,lags,train=False,plot=False):

  Data        = Data.drop(0)
  size        = len(Data)

  lagged_data = []

  for lag in range(1,lags+1):
    dummy_nan    = np.empty(197*93*lag)
    dummy_nan[:] = np.nan
    lagged_data.append(np.append(Data.Ugos.to_numpy().reshape((size, 1)),dummy_nan)[18321*lag:].reshape((size, 1)))
    lagged_data.append(np.append(Data.Vgos.to_numpy().reshape((size, 1)),dummy_nan)[18321*lag:].reshape((size, 1)))

  lagged_data = np.hstack(lagged_data)

  Set           = np.concatenate((

                          #Data.lat.to_numpy().reshape((size, 1)),
                          #Data.lon.to_numpy().reshape((size, 1)),
                          #LAT N
                          #Data.lat.to_numpy().reshape((size, 1))+0.25,
                          #LON W
                          #Data.lon.to_numpy().reshape((size, 1))-0.25,
                          #LAT S
                          #Data.lat.to_numpy().reshape((size, 1))-0.25,
                          #LON E
                          #Data.lon.to_numpy().reshape((size, 1))+0.25,
                          Data.ADT_C.to_numpy().reshape((size, 1)),
                          Data.ADT_N.to_numpy().reshape((size, 1)),
                          Data.ADT_E.to_numpy().reshape((size, 1)),
                          Data.ADT_W.to_numpy().reshape((size, 1)),
                          Data.ADT_S.to_numpy().reshape((size, 1)),
                          Data.V.to_numpy().reshape((size, 1)),
                          Data.U.to_numpy().reshape((size, 1)),
                          Data.Ugos.to_numpy().reshape((size, 1)),
                          Data.Vgos.to_numpy().reshape((size, 1)),

                          lagged_data
                        ),
                          axis=1)
  Set           = Set.reshape((size, Set.shape[1]))
  #if train==True:
  if plot == False:
    Set           = Set[~np.isnan(Set).any(axis=1)]
  if train == True:
    np.random.shuffle(Set)

  #del Data
  return Set

In [ ]:
Data_Train.head()

ADT_C   ADT_N   ADT_S   ADT_W   ADT_E    Ugos    Vgos       U       V
199  1.1360  1.1161  1.1506  1.1520  1.1170 -0.1647  0.2249 -0.1848  0.1925
200  1.1170  1.0992  1.1324  1.1360  1.0997 -0.1013  0.1679 -0.1819  0.2019
201  1.0997  1.0857  1.1150  1.1170  1.0842 -0.1006  0.1610 -0.1665  0.1700
202  1.0842  1.0716  1.0948  1.0997  1.0673 -0.0716  0.1544 -0.1305  0.1724
203  1.0673  1.0579  1.0731  1.0842  1.0504 -0.0682  0.1601 -0.0754  0.1805

In [ ]:
Train_set=generate_data(Data_Train)

In [ ]:
np.max(Train_set)

3.7229

In [ ]:
Test_set =generate_data(Data_Test,train=True)

In [ ]:
X_train = Train_set[:,:-10]
Y_train = Train_set[:,-10:]

In [ ]:
Train_set.shape

In [ ]:
data_joined = np.append(Train_set,Train_set,axis=0)

In [ ]:
# Define encoder model
inputs = tf.keras.layers.Input(shape=X_train.shape[1])
x = tf.keras.layers.Dense(5, activation='relu')(inputs)
encoded = tf.keras.layers.Dense(4, activation='relu')(x)
encoder_model = tf.keras.models.Model(inputs=inputs, outputs=encoded)

# Define decoder model
decoder_inputs = tf.keras.layers.Input(shape=(4,))
x = tf.keras.layers.Dense(5, activation='relu')(decoder_inputs)
decoded = tf.keras.layers.Dense(7, activation='relu')(x)
decoder_model = tf.keras.models.Model(inputs=decoder_inputs, outputs=decoded)

# Define autoencoder model
autoencoder_model = tf.keras.models.Model(inputs=inputs, outputs=decoder_model(encoder_model(inputs)))

# Compile autoencoder model
autoencoder_model.compile(loss='mse', optimizer='rmsprop')

# Train autoencoder model
autoencoder_model.fit(X_train, X_train, epochs=10, batch_size=10000)

Epoch 1/10
2900/2900 [==============================] - 18s 5ms/step - loss: 0.0220
Epoch 2/10
2900/2900 [==============================] - 15s 5ms/step - loss: 0.0084
Epoch 3/10
2900/2900 [==============================] - 15s 5ms/step - loss: 0.0084
Epoch 4/10
2900/2900 [==============================] - 15s 5ms/step - loss: 0.0084
Epoch 5/10
2900/2900 [==============================] - 15s 5ms/step - loss: 0.0084
Epoch 6/10
2900/2900 [==============================] - 15s 5ms/step - loss: 0.0084
Epoch 7/10
2900/2900 [==============================] - 15s 5ms/step - loss: 0.0084
Epoch 8/10
2900/2900 [==============================] - 14s 5ms/step - loss: 0.0084
Epoch 9/10
2900/2900 [==============================] - 14s 5ms/step - loss: 0.0084
Epoch 10/10
2900/2900 [==============================] - 14s 5ms/step - loss: 0.0084


In [ ]:
encoder_model.trainable = False

In [ ]:
# Define decoder model
inputs_encoder = tf.keras.layers.Input(shape=(4,))
x1 = tf.keras.layers.Dense(7, activation='relu')(inputs_encoder)
out = tf.keras.layers.Dense(10, activation='relu')(x1)
model_mlp = tf.keras.models.Model(inputs=inputs_encoder, outputs=out)

# Define autoencoder model
model = tf.keras.models.Model(inputs=inputs, outputs=model_mlp(encoder_model(inputs)))

#mae mse
model.compile(optimizer="Adamax",
              loss="huber_loss",#"mae",
              metrics=['RootMeanSquaredError',"mae",'acc'])

In [ ]:
np.sum(Y_train)

1579843.8393999944

In [ ]:
model.fit(X_train, Y_train, epochs=20, batch_size=10000)

Epoch 1/20
2900/2900 [==============================] - 26s 8ms/step - loss: 0.0381 - root_mean_squared_error: 0.2792 - mae: 0.1935 - acc: 0.0940
Epoch 2/20
2900/2900 [==============================] - 25s 8ms/step - loss: 0.0290 - root_mean_squared_error: 0.2413 - mae: 0.1731 - acc: 0.1174
Epoch 3/20
2900/2900 [==============================] - 24s 8ms/step - loss: 0.0288 - root_mean_squared_error: 0.2405 - mae: 0.1719 - acc: 0.1156
Epoch 4/20
2900/2900 [==============================] - 24s 8ms/step - loss: 0.0288 - root_mean_squared_error: 0.2405 - mae: 0.1719 - acc: 0.1156
Epoch 5/20
2900/2900 [==============================] - 25s 9ms/step - loss: 0.0288 - root_mean_squared_error: 0.2405 - mae: 0.1719 - acc: 0.1157
Epoch 6/20
2900/2900 [==============================] - 24s 8ms/step - loss: 0.0288 - root_mean_squared_error: 0.2405 - mae: 0.1719 - acc: 0.1156
Epoch 7/20
2900/2900 [==============================] - 24s 8ms/step - loss: 0.0288 - root_mean_squared_error: 0.2405 - mae:

log

---

* Epoch 1/10
905543/905543 [==============================] - 2086s 2ms/step - loss: 0.0918 - root_mean_squared_error: 0.1321 - acc: 0.3332
* Epoch 2/10
905543/905543 [==============================] - 1851s 2ms/step - loss: 0.0913 - root_mean_squared_error: 0.1314 - acc: 0.3376
* Epoch 3/10
905543/905543 [==============================] - 2245s 2ms/step - loss: 0.0910 - root_mean_squared_error: 0.1311 - acc: 0.3382
* Epoch 4/10
905543/905543 [==============================] - 1789s 2ms/step - loss: 0.0908 - root_mean_squared_error: 0.1308 - acc: 0.3383
* Epoch 5/10
905543/905543 [==============================] - 1658s 2ms/step - loss: 0.0905 - root_mean_squared_error: 0.1305 - acc: 0.3391
* Epoch 6/10
905543/905543 [==============================] - 1717s 2ms/step - loss: 0.0904 - root_mean_squared_error: 0.1303 - acc: 0.3391
* Epoch 7/10
905543/905543 [==============================] - 1797s 2ms/step - loss: 0.0902 - root_mean_squared_error: 0.1301 - acc: 0.3388
* Epoch 8/10
905543/905543 [==============================] - 1676s 2ms/step - loss: 0.0902 - root_mean_squared_error: 0.1300 - acc: 0.3388
* Epoch 9/10
905543/905543 [==============================] - 1650s 2ms/step - loss: 0.0901 - root_mean_squared_error: 0.1300 - acc: 0.3386
* Epoch 10/10
905543/905543 [==============================] - 1778s 2ms/step - loss: 0.0901 - root_mean_squared_error: 0.1299 - acc: 0.3385

In [ ]:
model.save('saved_model/model_7params_relu_8days(UV&ADT)_batch10000.h5')

In [ ]:
from keras.models import load_model
model = load_model('/content/model_7params_relu_5days(u_and_v).h5')

In [ ]:
#Test_set=generate_data(Data_Test,train=False,plot=False)
X_test=Test_set[:,:-16]
Y_test=Test_set[:,-16:]

In [ ]:
y_hat=model.predict(X_test,batch_size=1000)

12427/12427 [==============================] - 23s 2ms/step


perhitungan metrik

In [ ]:
# train
Train_set=generate_data(Data_Train,train=False,plot=False)
X_train=Train_set[:,:-10]
Y_train=Train_set[:,-10:]

In [ ]:
y_hat=model.predict(X_train,batch_size=1000)

14469/14469 [==============================] - 30s 2ms/step


In [ ]:
#korelasi u
print(np.corrcoef(y_hat[:,0],Y_test[:,0]))

In [ ]:
#korelasi v
print(np.corrcoef(y_hat[:,1],Y_test[:,1]))

In [ ]:
y_hat.shape

(14468576, 10)

In [ ]:
Y_train.shape

(14468576, 10)

In [ ]:
import torch
import torch.nn as nn
def velocity_rmse(u_sate,v_sate,u_pred,v_pred):
  u_obss = torch.tensor(u_sate, device='cpu')
  v_obss = torch.tensor(v_sate, device='cpu')
  u_hats = torch.tensor(u_pred, device='cpu')
  v_hats = torch.tensor(v_pred, device='cpu')

  def velocity(x,y):
    v = torch.sqrt(x**2+y**2)
    return v
  V_obss = velocity(u_obss,v_obss)
  V_hats = velocity(u_hats,v_hats)

  rmse = torch.sqrt(torch.mean((V_obss-V_hats)**2))
  nrmse= rmse/torch.std(V_obss)
  return rmse,nrmse

In [ ]:
for i in range(5):
  print(velocity_rmse(y_hat[2*i],y_hat[2*i+1],Y_train[2*i],Y_train[2*i+1]))


(tensor(0.0349, dtype=torch.float64), tensor(1.1805, dtype=torch.float64))
(tensor(0.0763, dtype=torch.float64), tensor(1.4961, dtype=torch.float64))
(tensor(0.0441, dtype=torch.float64), tensor(3.5746, dtype=torch.float64))
(tensor(0.0856, dtype=torch.float64), tensor(4.6238, dtype=torch.float64))
(tensor(0.0755, dtype=torch.float64), tensor(1.9032, dtype=torch.float64))


In [ ]:
np.mean(y_hat[:])

-0.000920115

In [ ]:
np.std(Y_test[:,0])# < 1.2 * 0.143

0.2764521522263844

In [ ]:
np.std(Y_test[:,1])# < 1.2 * 0.143

0.22018795028112498

In [ ]:
for i in range(8):
  print('u day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i],Y_test[:,2*i])[0,1]) )
  print('v day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i+1],Y_test[:,2*i+1])[0,1]))

NameError: ignored

In [ ]:
for i in range(8):
  print('u day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i],Y_train[:,2*i])[0,1]) )
  print('v day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i+1],Y_train[:,2*i+1])[0,1]))

u day1 0.9744387455212163
v day1 0.9634625978904815
u day2 0.9439371005553545
v day2 0.9232528982081721
u day3 0.9055757332286024
v day3 0.8724197072307812
u day4 0.8624272872224109
v day4 0.8148518968530294
u day5 0.8184322677038909
v day5 0.7558235195314363
u day6 0.776785096595545
v day6 0.6996107261720775
u day7 0.7394779393370978
v day7 0.649300293993116
u day8 0.7075591416655368
v day8 0.6064633736138533


In [ ]:
for i in range(8):
  print('u day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i],Y_train[:,2*i]+0.00001)[0,1]) )
  print('v day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i+1],Y_train[:,2*i+1]+0.00001)[0,1]))

/usr/local/lib/python3.9/dist-packages/numpy/lib/function_base.py:2829: RuntimeWarning: invalid value encountered in true_divide
  c /= stddev[:, None]
/usr/local/lib/python3.9/dist-packages/numpy/lib/function_base.py:2830: RuntimeWarning: invalid value encountered in true_divide
  c /= stddev[None, :]


u day1 nan
v day1 nan
u day2 0.8278622204050728
v day2 nan
u day3 nan
v day3 0.6874341140150425
u day4 nan
v day4 nan
u day5 nan
v day5 0.0005139198236582952


IndexError: ignored

In [ ]:
for i in range(5):
  print('u day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i],Y_test[:,2*i])[0,1]) )
  print('v day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i+1],Y_test[:,2*i+1])[0,1]))

NameError: ignored

In [ ]:
for i in range(5):
  print('u day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i],Y_test[:,2*i])[0,1]) )
  print('v day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i+1],Y_test[:,2*i+1])[0,1]))

u day1 0.9732341681253458
v day1 0.9645987041808459
u day2 0.9400688846546299
v day2 0.9227014178847522
u day3 0.8977181315866077
v day3 0.8689548829271466
u day4 0.8494921546652975
v day4 0.8072325621294497
u day5 0.7998578134145362
v day5 0.7428977246677257


In [ ]:
for i in range(5):
  print('u day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i],Y_train[:,2*i])[0,1]) )
  print('v day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i+1],Y_train[:,2*i+1])[0,1]))

u day1 0.001332244521816144
v day1 -0.005888907155451505
u day2 0.011012283926065168
v day2 -0.009433107809900322
u day3 0.013258622297402132
v day3 -0.008801611352118202
u day4 0.022595988834577387
v day4 -0.0072030609764500856
u day5 0.025650639320301982
v day5 -0.005060030471553175


In [ ]:
for i in range(5):
  print('u day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i],Y_test[:,2*i])[0,1]) )
  print('v day' + str(i+1) + ' ' + str(np.corrcoef(y_hat[:,2*i+1],Y_test[:,2*i+1])[0,1]))

u day1 0.9745240405925957
v day1 0.963865243896293
u day2 0.9440292965270493
v day2 0.9244793909997177
u day3 0.9056655086904937
v day3 0.874839930857565
u day4 0.8625843091966022
v day4 0.8185144746610207
u day5 0.8186189703035617
v day5 0.7603871011632038


In [ ]:
#RMSE u
print(np.sqrt(np.mean((y_hat[:,0]-Y_test[:,0])**2)))

In [ ]:
#RMSE v
print(np.sqrt(np.mean((y_hat[:,1]-Y_test[:,1])**2)))

rmse

In [ ]:
for i in range(8):
  print('u day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i  ]-Y_test[:,2*i  ])**2))))
  print('v day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i+1]-Y_test[:,2*i+1])**2))))

u day1 0.06468285192543193
v day1 0.059032484739611285
u day2 0.09381236164879572
v day2 0.08490901799379325
u day3 0.12055977247779016
v day3 0.10822553913048882
u day4 0.14361747597757
v day4 0.12857020365532593
u day5 0.16323873615700488
v day5 0.14542969375460266
u day6 0.17907076492272056
v day6 0.15889930924313908
u day7 0.19141902521025278
v day7 0.1692853188079687
u day8 0.20114785582104283
v day8 0.1770946925190422


In [ ]:
for i in range(5):
  print('u day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i  ]-Y_train[:,2*i  ])**2))))
  print('v day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i+1]-Y_train[:,2*i+1])**2))))

u day1 0.3503463049209796
v day1 0.34707056244077744
u day2 0.3381818650985837
v day2 0.3280522019957568
u day3 0.33198710966062633
v day3 0.3164675560142635
u day4 0.3242782251230942
v day4 0.30344948607816485
u day5 0.3187976780033523
v day5 0.29153066004489386


In [ ]:
for i in range(8):
  print('u day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i  ]-Y_train[:,2*i  ])**2))))
  print('v day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i+1]-Y_train[:,2*i+1])**2))))

u day1 0.28237268834000906
v day1 0.21878719784685935
u day2 0.18796513541240692
v day2 0.21878617839609996
u day3 0.28238583114149574
v day3 0.18044862270147138
u day4 0.28239371584585293
v day4 0.21878688705542476
u day5 0.28240190301478935
v day5 0.21878980831733189


IndexError: ignored

In [ ]:
for i in range(5):
  print('u day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i  ]-Y_train[:,2*i  ])**2))))
  print('v day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i+1]-Y_train[:,2*i+1])**2))))

u day1 0.3503463049209796
v day1 0.34707056244077744
u day2 0.3381818650985837
v day2 0.3280522019957568
u day3 0.33198710966062633
v day3 0.3164675560142635
u day4 0.3242782251230942
v day4 0.30344948607816485
u day5 0.3187976780033523
v day5 0.29153066004489386


In [ ]:
for i in range(5):
  print('u day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i  ]-Y_test[:,2*i  ])**2))))
  print('v day' + str(i+1) + ' ' + str(np.sqrt(np.mean((y_hat[:,2*i+1]-Y_test[:,2*i+1])**2))))

u day1 0.0645151360046761
v day1 0.06078160617047447
u day2 0.09581787603672876
v day2 0.08877521290853875
u day3 0.12383615980211506
v day3 0.11392995233775578
u day4 0.1483191947915314
v day4 0.13596484566338593
u day5 0.16873722265607735
v day5 0.15414708744433894


#import aviso

In [ ]:
AVISO_u  = xr.open_dataset('/content/drive/MyDrive/Tugas Akhir/check_u.nc')
AVISO_v  = xr.open_dataset('/content/drive/MyDrive/Tugas Akhir/check_v.nc')

In [ ]:
AVISO_v.longitude

In [ ]:
AVISO_u.latitude

In [ ]:
y_hat[-6*93*197,:].reshape(-1,93,197,10)

In [ ]:
124528580 % (93*197*10)

In [ ]:
plt.imshow(y_plot[1:,:].reshape(-1,93,197,10)[0,:,:,0])

In [ ]:
plt.imshow(AVISO_u.ugos[0,:,:].to_numpy().reshape(1,-1).reshape(88,192))

#plot